statev2/state/src/applicator/order_book.rs

```diff
@@ -0,0 +1,488 @@
+        let prev_nullifier = order_info.public_share_nullifier;
+        let new_nullifier = proof.reblind_proof.statement.original_shares_nullifier;
+        if prev_nullifier != new_nullifier {
+            Self::update_order_nullifier(order_id, prev_nullifier, new_nullifier, tx)?;
+        }
+
+        // Update the order's info
+        order_info.state = NetworkOrderState::Verified;
+        order_info.public_share_nullifier = proof.reblind_proof.statement.original_shares_nullifier;
+        order_info.validity_proofs = Some(proof);
+        Self::write_order_info(&order_info, tx)
+    }
+
+    /// Cancel an order
+    fn cancel_order_with_tx(&self, order_id: &OrderIdentifier, tx: &DbTxn<'_, RW>) -> Result<()> {
+        let mut order = Self::read_order_info(order_id, tx)?;
+        order.state = NetworkOrderState::Cancelled;
+        order.validity_proof_witnesses = None;
+        order.validity_proofs = None;
+
+        Self::write_order_info(&order, tx)?;
+        self.system_bus().publish(
+            ORDER_STATE_CHANGE_TOPIC.to_string(),
+            SystemBusMessage::OrderStateChange { order },
+        );
+
+        Ok(())
+    }
+
+    // ----------------------
+    // | Order Info Helpers |
+    // ----------------------
+
+    /// Reads the order info for the given order from storage
+    ///
+    /// Errors if the order is not present
+    fn read_order_info<T: TransactionKind>(
+        order_id: &OrderIdentifier,
+        tx: &DbTxn<'_, T>,
+    ) -> Result<NetworkOrder> {
+        let order_key = Self::order_key(order_id);
+        tx.read(ORDERS_TABLE, &order_key)
+            .map_err(StateApplicatorError::Storage)?
+            .ok_or_else(|| StateApplicatorError::MissingEntry(order_key))
+    }
+
+    /// Writes the order info for the given order to storage
+    fn write_order_info(order: &NetworkOrder, tx: &DbTxn<'_, RW>) -> Result<()> {
+        let order_key = Self::order_key(&order.id);
+        tx.write(ORDERS_TABLE, &order_key, order)
+            .map_err(StateApplicatorError::Storage)
+    }
+
+    // --------------------------
+    // | Order Priority Helpers |
+    // --------------------------
+
+    /// Get the priority of a cluster
+    fn get_cluster_priority_with_tx<T: TransactionKind>(
+        cluster_id: &ClusterId,
+        tx: &DbTxn<'_, T>,
+    ) -> Result<u32> {
+        tx.read(PRIORITIES_TABLE, cluster_id)
+            .map_err(StateApplicatorError::Storage)
+            .map(|priority| priority.unwrap_or(CLUSTER_DEFAULT_PRIORITY))
+    }
+
+    /// Write an order priority to the DB
+    fn write_order_priority_with_tx(order: &NetworkOrder, tx: &DbTxn<'_, RW>) -> Result<()> {
+        // Lookup the cluster priority and write the order's priority
+        let cluster_priority = Self::get_cluster_priority_with_tx(&order.cluster, tx)?;
+        let priority = OrderPriority {
+            cluster_priority,
+            order_priority: ORDER_DEFAULT_PRIORITY,
```

Comment:
It feels a bit strange to me that this function overwrites the `OrderPriority.order_priority` field. The function signature / docstring doesn't really indicate this "reset to default" behavior, nor does it indicate that the `NetworkOrder` is expected to represent an order with an uninitialized priority. Perhaps it would make sense to add an `order_priority` field to the `NetworkOrder` type?

Reply:
This method is only called when a new order is added, so no priority events have occurred that would change the order's priority.

Plus I'd rather not form consensus over priorities, I'd rather optimistically gossip (if anything at all) about protocol faults that would lower a priority. This is far less complex